### ISSUE TITLE
Dark Mode: CSS Declarations hard to read in Editor

### ISSUE THREAD

**Issue by mstellaluna** [freeCodeCamp/freeCodeCamp#16319]
#### Challenge Name
Basic CSS: Style Multiple Elements with a CSS Class and beyoned

#### Issue Description

The CSS declarations in dark mode are hard to read. They are somethings navy blue on black.
This also happens for anchor tags in the HTML section.

#### Browser Information

Browser Name, Version:
Operating System: FireFox 57.0 (64-bit) and Chrome Version 63.0.3239.84 (Official Build) (64-bit)
Mobile, Desktop, or Tablet: Laptop Windows 10 Professional 64-bit

#### Your Code

```js



```
#### Screenshot

![image](https://user-images.githubusercontent.com/22984739/34328638-4162073a-e8b2-11e7-9f48-a2578dc052e4.png)

**Comment by ncaron** [comment]
@mstellaluna Is it possible that changes in a different PR may have affected this? I do see exactly that on the live beta site but not on my up-to-date staging branch. I don't see any code highlighting at all actually, which is a bigger issue in my opinion.

![no-highlight](https://user-images.githubusercontent.com/13804601/34345766-d78d8d40-e9c7-11e7-82e7-ea759b35265c.png)

**Comment by mstellaluna** [comment]
@ncaron not sure, I am noticing a large amount of issues with dark mode. I haven't opened new issues for them yet ..

**Comment by profoundhub** [comment]
It would be nice to have a switch so that anyone can activate `dark mode`, even when not logged in.

**Comment by ncaron** [comment]
Perhaps having a switch in the header near the search bar might be good. We would be able to switch when not logged in but also switch anytime without having to leave our challenge to do so.

**Comment by systimotic** [comment]
@ncaron I've also had issues with the syntax highlighting for months locally, but it has always worked fine on the beta (on Windows, both in Chrome and Firefox). I've seen various reports of it working/not working for others. I looked into it a little bit today, and saw a report that it has something to do with webpack (https://github.com/JedWatson/react-codemirror/issues/34). This would explain why there's a difference between the development environment and the built, live beta.

---

I found the styles for the night mode here: https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/challenges.less#L96
We could manually tweak these styles, or use an existing CodeMirror theme: https://github.com/codemirror/CodeMirror/blob/master/theme/dracula.css

**Comment by raisedadead** [comment]
We should just use the `default` code mirror theme for day mode and `dracula theme` for for night mode in code mirror config.

Its well maintained for very many coding languages.

**Comment by huyenltnguyen** [comment]
@systimotic Thanks for all the info, it helps a lot.
I've made some progress on this issue and I'm quite positive that #16284 is related to this one (see the screenshot below). I've found that the problem is actually in some of the React components (`Editor.jsx` and `Code-Mirror-Skeleton.jsx`), in which we didn't add the theme correctly. And even if we did, the theme wouldn't display because we didn't load/import their CSS into our stylesheet.
The thing is, the changes I made here were on the development version. Since the built version doesn't have those `.jsx` files, I suppose overriding `codemirror` style as @systimotic suggested is a quick fix then.

![nightmode](https://user-images.githubusercontent.com/25715018/34453445-6081dc1e-ed88-11e7-9394-3db791dcc74a.png)

**Comment by huyenltnguyen** [comment]
Hi @raisedadead, sorry for bothering you, but since you're more familiar with codemirror, I think you're the right person to ask.
I think I was on the wrong branch when I commented, in the comment above, that I didn't see some React components in the codebase (not sure where my head was at at the time).
Anyway, I see the class `cm-s-monokai` in `Code-Mirror-Skeleton.jsx`, from which I assume we originally wanted to use `monokai` theme, but codemirror doesn't recognize it and uses the `default` theme as a fallback.
https://github.com/freeCodeCamp/freeCodeCamp/blob/9263b91f73b292db2a8d0b141b574eda0296e699/common/app/routes/Challenges/Code-Mirror-Skeleton.jsx#L34
Now, my solution, after reading the [codemirror doc](https://codemirror.net/doc/manual.html), is:
- Remove the `cm-s-monokai` class in `Code-Mirror-Skeleton.jsx`.
- Add theme name to the config object in [Editor.jsx](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/views/classic/Editor.jsx#L22).
- Import theme's style into our stylesheet.

The result can be seen in the screenshot in my comment above. We don't have syntax highlighting in local/development version because apparently `react-codemirror` and `webpack` don't like each other, but from the screenshot, the Editor component now has `cm-s-dracula` class and its style as well (dark blue background and white cursor).

Here are my questions:
- Does this look legit?
- If so, and now that we've found our long lost `monokai`, should we just use it on both day and night modes (the official FCC doesn't have night mode, but uses `monokai` for day mode)? Or should we ditch it and use `default` for day and `dracula` for night (that's a bit mean, but who am I to judge)?
- We have 2 `Editor.jsx` files: a [classic]( freeCodeCamp/common/app/routes/Challenges/views/classic/Editor.jsx ) one and a [modern](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/views/Modern/Editor.jsx) one. I was working on the `classic` one, but am not sure how the files are used and if I should apply the change on both of them.
- Will my PR be approved if I can only test it partly because of the mentioned syntax highlighting issue? I totally understand if it won't, though.

Sorry for the wall of text.
Thank you in advance. Have a nice weekend!

**Comment by raisedadead** [comment]
> Hi @raisedadead, 
Hey there! @huyenltnguyen , how is going? 

> sorry for bothering you, but since you're more familiar with codemirror, I think you're the right person to ask.

Hit me up any time, I may not be an expert though. And thanks a lot for taking the time and efforts in investigating and helping.

Okay now coming back to the queries:

Yes, you are correct. The styles which are imported in `main.less` (entry point for all things css usually) may not be consumed by the new react components at all places.

1. There are four areas where the code mirror is in play of which one is a dummy (the skeleton, where its not even used as a component). 
    These areas where CSS should be consumed are : [the classic editor](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/views/classic/Editor.jsx) (for the standard challenges), [the mordern editor](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/views/Modern/Editor.jsx) (for the react challenges, me thinks), [the console output](common/app/routes/Challenges/Output.jsx)
2. All of the these above use the [`challenges.less`
](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/challenges.less) as the entry point to the styles.

Hope the above helps you in debugging further. 
You should not have to make any changes to the skeleton and while it needs improvement, its a separate issue all-together #15700 and can be taken in parallel  to this issue.

Now,

Here are my questions:

> - Does this look legit?

I can give an accurate feedback once we have more information and a PR to test.

> - If so, and now that we've found our long lost monokai, should we just use it on both day and night modes (the official FCC doesn't have night mode, but uses monokai for day mode)? Or should we ditch it and use default for day and dracula for night (that's a bit mean, but who am I to judge)?

Yup, IMHO, we would prefer the `default` for the day and `dracula` works amazing for night.

> - We have 2 Editor.jsx files: a classic one and a modern one. I was working on the classic one, but am not sure how the files are used and if I should apply the change on both of them.

Yes, both places. See above for why.

> - Will my PR be approved if I can only test it partly because of the mentioned syntax highlighting issue? I totally understand if it won't, though.

Totally yes, any improvement is good as long as we are tracking it. So, if you are able to address a issue partially with PR. We just need sure a new one is opened when yours is merged in, to track additional work.

If you prefer to keep the original open instead to continue discussion, then you can choose to do so by using `refers #xxxxx` instead of closes `#xxxxx` in the PR. You can ignore the issue no is commit messages in that case. 

Have a nice weekend too and happy contributing.!

**Comment by raisedadead** [comment]
FYI we also want to upgrade to react-codemirror2 as being discussed in the #16397 and avoid the webpack issue that @systimotic is refering to which I am unaware of.

/cc @BerkeleyTrue even more reason..! lol.

**Comment by huyenltnguyen** [comment]
Hi @raisedadead, thanks so much for the reply, it does help me in debugging and understanding the codebase a bit more.
I have some updates and would really appreciate your input and your help.

So this is my to-do list:
- [x] Import the `dracula` style into [challenges.less](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/challenges.less).
- [ ] Add logic to both of the `Editor.jsx` files to toggle theme based on the day/night mode.
- [ ] Change console output background based on the day/night mode.

I'm stuck at the second item and could use some help here. To be honest, the amount(ain) of React/Redux and theirs related libraries is very daunting to me. I know that I need to get the user selected mode (day/night, or `default/night` in the code). The [client/night-mode-epic.js file](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/client/epics/night-mode-epic.js) points me to the [common/app/redux/index.js file](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/redux/index.js) and the [common/app/entities/index.js](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/entities/index.js) and I got lost from there.

Onto the console. I'm not really stuck as I know how to change its style, I just need your opinion on this, though. Our console was given the theme `monokai`, however, it doesn't have the theme's style because we didn't import it (the style) into `challenges.less` (I can tell because the console has white background while monokai's is black). IMHO, setting theme for the console is unnecessary since we only display some text and don't need any syntax hightlighting. My suggestion is to remove the `monokai` theme in the console config and add some style for its night mode instead. What do you think?
https://github.com/freeCodeCamp/freeCodeCamp/blob/6521839ff45561acbd497314c8455afb2d9a40d4/common/app/routes/Challenges/Output.jsx#L14

Also another thing, in the [classic.less file](https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/common/app/routes/Challenges/views/classic/classic.less), the `.night {}` part is a duplicate of `challenges.less`'s and I think it's safe to remove that part.

**Comment by raisedadead** [comment]
> I'm stuck at the second item ...

I'll have to get back to you on this after taking a deeper look at the implementation, @BerkeleyTrue or @Bouncey are the best persons to seek guidance though. 

The need for storing the theme on UserObject was to be able to sync it as a preference across  sessions.

That said, it should not have to be too complicated. Bear with us while we get back to you.

> Onto the console. ...

Agreed with your suggestion, but its worth keeping the consistency as with `default` and `dracula` switch. But, I'll let you decide which ever is simpler to implement (i.e adding custom style or toggling themes, in your snippet for `Output.jsx` above).

> Also another thing, in the classic.less file ...

Agreed.

**Comment by BerkeleyTrue** [comment]
@huyenltnguyen You'll need to grab the themeSelector in app/redux and add it to mapStateToProps. Then pass this through to the editors.

**Comment by huyenltnguyen** [comment]
@raisedadead & @BerkeleyTrue Thanks for the direction. I'll take a look at them.

**Comment by huyenltnguyen** [comment]
@BerkeleyTrue, I have a dumb question here. When you said I needed to pass the theme to the editors, did you mean passing it as a React prop or passing it in some kind of a Redux way? I tried implementing the former and found myself passing it 5-6 levels down. It feels a bit weird to me, so I decided to seek clarification first.

**Comment by BerkeleyTrue** [comment]
> found myself passing it 5-6 levels down

What do you mean by this?

**Comment by huyenltnguyen** [comment]
@BerkeleyTrue In [common/app/App,jsx](https://github.com/freeCodeCamp/freeCodeCamp/blob/68c4485e12ecc5876e5ce713e817ea2727a05b8f/common/app/App.jsx), I did this:
```
import {
...
themeSelector
} from './redux';

const mapStateToProps = state => {
  ...
  const theme = themeSelector(state);
  return {
    ...
    theme
  }
}

```
Then I started passing `theme` to the Child component as a prop, and used React Dev Tool to track and kept passing the theme down the tree eventually to the Editor components. 
https://github.com/freeCodeCamp/freeCodeCamp/blob/68c4485e12ecc5876e5ce713e817ea2727a05b8f/common/app/App.jsx#L70

**Comment by BerkeleyTrue** [comment]
There is no reason you need to import the themeselector at the top component. You can import it in the editor component directly.

**Comment by huyenltnguyen** [comment]
Got it. Thanks, @BerkeleyTrue.

**Comment by huyenltnguyen** [comment]
Hi @raisedadead, I know you're very busy, but I just want to give you an update.
I just checked the live beta, the dracula theme is available in night mode now, and it looks so beautiful!
However, you were right about the console. Its night mode display is horrible (screenshot below) and I should've added the logic there as well instead of just manually changing its background and text color. (In my defense, I didn't know it uses syntax highlighting too, all I see on my local machine is white text on dark background lol).
Anyway, a PR is on its way. It should close this issue as well.

![untitled](https://user-images.githubusercontent.com/25715018/35193150-65b259de-fed1-11e7-9468-6f937f0a0a59.png)

**Comment by raisedadead** [comment]
@huyenltnguyen Sure, looking forward to it.. thanks for working on it.
